crates/typst/src/layout/grid/layout.rs

```diff
@@ -278,6 +373,114 @@ impl CellGrid {
+                ))
+            })
+    };
+    match (cell_x, cell_y) {
+        // Fully automatic cell positioning. The cell did not
+        // request a coordinate.
+        (Smart::Auto, Smart::Auto) => {
+            // Let's find the first available position starting from the
+            // automatic position counter, searching in row-major order.
+            // We have to sum 'next_auto_index' at the end since we'd have
+            // the position after the initial skip, not from the start of the
+            // vector.
+            let next_auto_index = *auto_index;
+            let resolved_index = resolved_cells
+                .iter()
+                .skip(next_auto_index)
+                .position(Option::is_none)
+                .map(|i| i + next_auto_index)
+                .unwrap_or_else(|| resolved_cells.len());
+
+            // Ensure the next cell with automatic position will be
+            // placed after this one (maybe not immediately after).
+            *auto_index = resolved_index + 1;
+
+            Ok(resolved_index)
+        }
+        // Cell has chosen its exact position.
+        (Smart::Custom(cell_x), Smart::Custom(cell_y)) => {
+            if cell_x >= columns {
+                return Err(HintedString::from(eco_format!(
+                    "a cell could not be placed at invalid column {cell_x}"
+                )));
+            }
+            cell_index(cell_x, cell_y)
+        }
+        // Cell has only chosen its column, not its row.
+        (Smart::Custom(cell_x), Smart::Auto) => {
+            if cell_x >= columns {
+                return Err(HintedString::from(eco_format!(
+                    "a cell could not be placed at invalid column {cell_x}"
+                )));
+            }
+            // Let's find the first row which has that column available.
+            let mut resolved_y = 0;
+            while let Some(Some(_)) = resolved_cells.get(cell_index(cell_x, resolved_y)?)
+            {
+                // Try each row until either we reach an absent position
+                // (`Some(None)`) or an out of bounds position (`None`),
+                // in which case we'd create a new row to place this cell in.
+                resolved_y += 1;
+            }
+            cell_index(cell_x, resolved_y)
+        }
+        // Cell has only chosen its row, not its column.
+        (Smart::Auto, Smart::Custom(cell_y)) => {
+            // Let's find the first available column in the chosen row.
+            // The first available position after the first cell of the row,
+            // within up to 'columns' positions, will be the first available
+            // column. We chain extra absent positions (None) in case the
+            // vector isn't large enough yet to hold all cells in the row
+            // (in which case we can place the cell at the first out of bounds
+            // position, since it's available).
+            resolved_cells
```

Comment:
I feel like we should probably either use loops consistently or iterators in this function. The first and last are iterators, but the middle one a loop. 

This final one here is a bit tricky to understand. Would something simpler like this below work?
```rust
let start = cell_index(0, cell_y)?;
(start..start + columns)
  .find(|i| !matches!(resolved_cells.get(i), Some(_)))
``` 
Generally, I think directly iterating over the candidate indices might sometimes be simpler than using position() and than mapping over an addition of a base index.

Reply:
I used to use exclusively loops before, but I thought that it didn't make much sense to do something like `.get(0)`, `.get(1)` etc. when that's precisely iterating over the vector after a certain point (through `.skip`) and calculating the position. (It was especially jarring in the case with `auto_index`, where I'd just add 1 on every iteration, which felt like a code smell :p)  And Rust's standard library probably does this job better than us.

Regarding the final iterator, I had basically that setup you propose, but then I changed to the current one at e10200102a6e478fbed370ce78f190eb42c94d8d due to the reasons above, and it basically works by iterating over the whole row. We can do that since the vector is in row-major order. It's something like `&vec[first_index..(first_index + columns)].iter().position(...)` which will give you the first available column, except that I can't use a slice cuz `first_index + columns` might be out of bounds. If we push full rows, we could use slices, but AFAIK iterators are still faster (at least this [link](https://stackoverflow.com/questions/32690678/is-it-more-efficient-to-slice-an-array-or-use-iteratorskip) suggested so based on some ad-hoc and totally unscientific research).

But the middle one, as far as I'm aware, has to be done with loops since I don't think we can tell iterators to skip `c` elements on each iteration in a very practical way. I considered using `chunks()`, but it didn't feel much better than just using the loop like that.

Reply:
With that said, we can probably experiment with returning to loop-based approaches and seeing how they perform.

Reply:
My guess would be that this code will be far from being the bottleneck of a Typst execution, so iterator vs range + get will probably have no observable difference.

Reply:
Yeah fair enough. Though the `auto_index` case would still worry me. Either way, I switched back to loops now and performance seems to be roughly the same as before, so we should be fine.